persist/fs/files.go

```diff
@@ -324,20 +472,91 @@ func findSubDirectoriesAndPaths(directoryPath string) (directoryNamesToPaths, er
 	return subDirectoriesToPaths, nil
 }
 
-func filesetFiles(filePathPrefix string, namespace ident.ID, shard uint32, pattern string) ([]string, error) {
-	shardDir := ShardDirPath(filePathPrefix, namespace, shard)
-	return findFiles(shardDir, pattern, func(files []string) sort.Interface {
+func snapshotFiles(filePathPrefix string, namespace ident.ID, shard uint32, pattern string) ([]SnapshotFile, error) {
+	shardDir := ShardSnapshotsDirPath(filePathPrefix, namespace, shard)
+	byTimeAsc, err := findFiles(shardDir, pattern, func(files []string) sort.Interface {
+		return snapshotsByTimeAndIndexAscending(files)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		latestBlockStart   time.Time
+		latestIndex        int
+		latestSnapshotFile SnapshotFile
+		snapshotFiles      = []SnapshotFile{}
+	)
+	for _, file := range byTimeAsc {
+		currentFileBlockStart, index, err := TimeAndIndexFromSnapshotFilename(file)
+		if err != nil {
+			return nil, err
+		}
+
+		if latestBlockStart.IsZero() {
+			latestSnapshotFile = NewSnapshotFile(currentFileBlockStart)
+			latestIndex = index
+		} else if !currentFileBlockStart.Equal(latestBlockStart) || latestIndex != index {
+			snapshotFiles = append(snapshotFiles, latestSnapshotFile)
+			latestSnapshotFile = NewSnapshotFile(currentFileBlockStart)
+			latestIndex = index
+		}
+		latestBlockStart = currentFileBlockStart
+
+		latestSnapshotFile.Index = index
+		latestSnapshotFile.AbsoluteFilepaths = append(latestSnapshotFile.AbsoluteFilepaths, file)
+	}
+	snapshotFiles = append(snapshotFiles, latestSnapshotFile)
```

Comment:
is there a nil check missing here? i.e. this returns a slice with a single element if there are no snapshot files, should you be returning an empty slice instead?

could you add a test for this too

Reply:
good catch, thanks. Fixed